discovery-protocols - Refactor to Modules\Links

Npeca75 opened the pull request:
fix #17873 

1. split modules per OS
2. centralized discover_new_device by hostname
3. centralized discover_new_device by ip (if applicable)
4. lldp v1
5. lldp v2
6. filter out Mikrotik parrent interface in port descr. EX: br0/ether1 -> ether1 for proper port matching
7. filter out LLDP artefacts from jetstream devices where switch send LLDP capabilities embeded in host name
8. handle old T1600G-28TS 2.0 switch



DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

Npeca75 committed:
Module\Links

PipoCanaja commented:
Hi @Npeca75 
Seems that discovery-protocols is broken on VRP (huawei) devices with this PR. Will dig to find out why.

Npeca75 commented:
> Will dig to find out why, and fix it hopefully.

thank you

PipoCanaja commented:
It does indeed. I reordered because in fact ifName is the only compliant way according to MIB. So this one should be the first tried, and the others would be for bad LLDP implementations. 
Will submit a new file soon.

PipoCanaja committed:
reorder local-port discovery

PipoCanaja committed:
Update Links.php

PipoCanaja committed:
style

PipoCanaja committed:
revert VRP tests

Npeca75 committed:
remote port name

Npeca75 committed:
revert dell-sonic

PipoCanaja commented:
I'll fix the 'panos' hex2ascii as well

Npeca75 committed:
edgeswitch_usw-flex-xg

Npeca75 committed:
dot1dBasePortIfIndex

PipoCanaja committed:
Fix Panos + is_null && empty

PipoCanaja committed:
style

PipoCanaja committed:
style

PipoCanaja commented on tests/data/eltex-mes23xx_mes2324p.json:
```diff
@@ -21425,7 +21425,7 @@
                     "active": 1,
                     "protocol": "lldp",
                     "remote_hostname": "AVRIP",
-                    "remote_port": "port e0/1/1 (00017afdb21c)",
```
The interface name is lost here it seems.

PipoCanaja commented on tests/data/eltex-mes23xx_mes2348p.json:
```diff
@@ -16785,7 +16785,7 @@
                     "active": 1,
                     "protocol": "lldp",
                     "remote_hostname": "bt-stadion-payload-ap-07",
-                    "remote_port": "bridge1/vlan75-payload (2cc81b5ba251)",
```
Seems that interfaces are mixed up here

Npeca75 replied:
let's wait for test to reach Eltex OS
now with reversed local_port_id find order

Npeca75 commented:
@PipoCanaja 
could you look at
PHP Error(2): Undefined array key 0 in /home/runner/work/librenms/librenms/includes/polling/functions.inc.php on line 204 ?
i tried to fix it, but still ...

Npeca75 committed:
lldpRemPortIdSubtype => portComponent
mes2324p

Npeca75 commented:
> Where do you see that one ? Don't have it here.

in workflow run / test

Npeca75 commented:
ok @PipoCanaja 
eltex-mes23xx_mes2324p.json

```
  "lldpRemPortIdSubtype" => "portComponent"     <---- this Subtype is not covered with code, working on it
  "lldpRemPortId" => "port e0/1/1"
```

Npeca75 commented:
> Where do you see that one ? Don't have it here.

since @murrant enabled error reporting in #17672  ... workflow/test window is full of garbage :)

PipoCanaja commented:
> > Where do you see that one ? Don't have it here.
> 
> since @murrant enabled error reporting in #17672 ... workflow/test window is full of garbage :)

Got it, opened a PR to try to fix a few of these.

Npeca75 commented:
> Got it, opened a PR to try to fix a few of these.

as we work on this PR, i started to fix them, please look at modified files in this PR

PipoCanaja commented:
Please avoid mixing other issues in this PR. Let's keep it for discovery-protocols.

PipoCanaja commented:
@librenms/reviewers  : Should we rename this module ?

Npeca75 commented:
@PipoCanaja 
we are good to go ...

Npeca75 commented:
> Let's keep it for discovery-protocols.

ok, will revert

PipoCanaja commented on tests/data/aos7.json:
```diff
@@ -78125,519 +78125,2135 @@
+                    "remote_platform": null,
+                    "remote_version": "",
+                    "ifAlias": "Alcatel-Lucent OS6860 GNI 2/1/26",
+                    "ifDescr": "Alcatel-Lucent OS6860 GNI 2/1/26",
```
Are we sure about these ? it seems that we have a lot more LLDP entries, and some are redundant.

Npeca75 replied:
hi @PipoCanaja 
spun off AOS7 snmpsim, and observed the resulting DB

it is 200++ port device, big one
and there is total 68 links detected on AOS7 test data

observing DB, there is no duplicate !!!
every 
`local_port_id + remote_hostname + remote_port` 
combination is uniq !

simple visual exploring of 'LLDP-MIB::lldpRemTable' will show that there is much much more entry than in old test.json
have no idea why old test.json was so small

looked at the old test data, every entry from old test is included in new discovery data
so, at least this part is OK

PipoCanaja commented:
> > Let's keep it for discovery-protocols.
> 
> ok, will revert

Thx. Feel free to add your fix to the #17784

Npeca75 commented:
> links as it is a bit more descriptive

i choose `Links` because it is the reality: we have links between devices

and another thing, how to shorten `discovery-protocols` ?
you will end with something like `function discoverDiscoveryProtocols` or similar, too much discovery in one line :)
please live it as `Links` if it is possible

murrant commented:
find_device_id and find_port_id will need to be moved at some point.  (also, they are kind of a mess)   Maybe it would be better to rewrite those at a future time.

Npeca75 committed:
LinksDiscovery -> LinkDiscovery

PipoCanaja committed:
fix regexp

We want to match only HEX encoded MacAddresses, so 5 times :3a:, which is 5 times `:` after conversion

murrant commented:
I don't think PortCache is the correct place for that code.

Npeca75 commented:
> I don't think PortCache is the correct place for that code.

ok, any suggestion ? or help ?
i really like to finish this PR

Npeca75 commented:
closed by mistake

murrant commented:
I initially said they shouldn't be moved in the PR.  Sorry, I was kind of making a mental note.  I didn't mean to have you move the code around.

If I had to fix them, I would delete them and write something new.  So, just moving them around in the code is not very effective.

Npeca75 commented:
> This HEX conversion

hi @PipoCanaja 

well, it is OK now, but i am unsure about this
```
&& strlen($tmpPortId) != 12) { // do not convert MAC
```
should we try to avoid false converting MAC to ascii ? or blindly convert ?
after all, `$data['lldpRemPortIdSubtype'] == 'interfaceName'` should filter out MAC addresses, but still ...
please rewrote as you wish

PipoCanaja commented:
> > This HEX conversion
> 
> hi @PipoCanaja
> 
> well, it is OK now, but i am unsure about this
> 
> ```
> && strlen($tmpPortId) != 12) { // do not convert MAC
> ```
> 
> should we try to avoid false converting MAC to ascii ? or blindly convert ? after all, `$data['lldpRemPortIdSubtype'] == 'interfaceName'` should filter out MAC addresses, but still ... please rewrite as you wish

Went for the blind version. If an interface has the exact same length as a MAC, we would be breaking things :D

PipoCanaja reviewed (dismissed):
Looks good. @murrant I'll leave it to you to merge the PR

sorano commented:
Tested this further and with the refactor code applied then the procurve is not displaying duplicated LLDP neighbours in webui.

And testing on Cisco if I disable CDP globally with "no cdp run" then the webui shows lldp neighbours properly without duplication, but if both CDP and LLDP is enabled then no LLDP neighbours are displayed in webui at all.

murrant commented:
FYI, I'm going to send a PR to fix up/refactor find_device_id and find_port_id.

murrant commented:
After that PR, we will start looking at what it takes to get this merged. Thanks for your patience.

Npeca75 committed:
merge discovery methods

murrant reviewed (changes requested):
This has a lot of work left, please hold until I can further review.

Npeca75 committed:
/opt/librenms/tests/data/nxos_n3k-3064pq.json

Npeca75 committed:
Merge branch 'master' into lldp

Jellyfrog commented:
@Npeca75 please check the changed test data in https://github.com/librenms/librenms/pull/18036
A lot of invalid MAC addresses …

Jellyfrog commented on LibreNMS/Modules/Links.php:
```diff
@@ -0,0 +1,317 @@
+                ! empty($ip) &&
+                LibrenmsConfig::get('discovery_by_ip', false) &&
+                LibrenmsConfig::get('autodiscovery.xdp') === true) { //name lookup failed, try with IP
+                $link->remote_device_id = discover_new_device(
```
Can this code even call these legacy functions ?

Npeca75 replied:
hi @Jellyfrog 

yes, it is working, tested many times

murrant replied:
To add a little extra. I greatly appreciate contributions.   If you want things merged faster, make them easier to merge.

Npeca75 committed:
Merge branch 'master' into lldp

Npeca75 commented:
> Avaya/Extreme

Could you upload somewhere comple snmp walk?
snmpwalk -Oneb -v2c -c XXXX 192.168.0.x . > /tmp/out

Npeca75 committed:
Merge branch 'master' into lldp

Npeca75 committed:
Merge branch 'master' of https://github.com/librenms/librenms into lldp

Npeca75 committed:
populate remManAddr

jayib commented:
All of my tests reveal that it works now.  I am able to discover devices properly

laf reviewed (changes requested):
I think only that update to the test unit from my side.

@murrant Thoughts on this?

Npeca75 committed:
Merge branch 'master' of https://github.com/librenms/librenms into lldp

Npeca75 committed:
Descr -> Desc

Npeca75 committed:
Merge branch 'master' of https://github.com/librenms/librenms into lldp

Npeca75 committed:
MacUtilTest

Npeca75 committed:
fs-switch

Npeca75 commented on LibreNMS/Util/Mac.php:
```diff
@@ -62,6 +62,7 @@ public function __construct(string $mac)
      *  001234-ABCDEF
      *  0012.34AB.CDEF
      *  00:02:04:0B:0D:0F
+     *  00 02 04 0B 0D 0F
```
> Can you update MacUtilTest.php please to include this.

done

Npeca75 committed:
Merge branch 'master' of https://github.com/librenms/librenms into lldp

Npeca75 committed:
rector

Npeca75 committed:
Merge branch 'master' of https://github.com/librenms/librenms into lldp

Npeca75 committed:
Merge branch 'master' of https://github.com/librenms/librenms into lldp

Npeca75 committed:
styleci/rector

Npeca75 committed:
Merge branch 'master' of https://github.com/librenms/librenms into lldp

Npeca75 committed:
lint

Npeca75 committed:
Merge branch 'master' of https://github.com/librenms/librenms into lldp

Npeca75 committed:
ci

Npeca75 committed:
fsbdcom

Npeca75 committed:
Merge branch 'master' of https://github.com/librenms/librenms into lldp

Npeca75 committed:
test + ci

Npeca75 committed:
bdcom

Npeca75 committed:
Merge branch 'master' into lldp

Npeca75 committed:
lint

Npeca75 committed:
Polling\ConnectivityHelper

Npeca75 committed:
Merge branch 'master' of https://github.com/librenms/librenms into lldp

Npeca75 committed:
mikrotik cdp/mndp
